xocto/ranges.py

```diff
@@ -944,3 +946,56 @@ def any_overlapping(ranges: Iterable[Range[T]]) -> bool:
             return True
         range_set.add(range)
     return False
+
+
+def as_finite_datetime_periods(
+    periods: Iterable[HalfFiniteDatetimeRange | DatetimeRange],
+) -> Sequence[FiniteDatetimeRange]:
+    """
+    Casts the given date/time periods as finite periods.
+
+    This is useful when working with potentially infinite ranges that are
+    known to be finite e.g. due to intersection with a finite range.
+
+    Raises:
+        ValueError: If one or more periods is not finite.
+    """
+    finite_periods = []
+
+    for period in periods:
+        if period.start is None or period.end is None:
+            raise ValueError("Period is not finite at start or end or both")
+
+        finite_periods += [FiniteDatetimeRange(period.start, period.end)]
+
+    return finite_periods
+
+
+def as_date_periods(
+    periods: Iterable[FiniteDatetimeRange],
+    *,
+    date_converter: Callable[[datetime.datetime], datetime.date],
+) -> Sequence[FiniteDateRange]:
+    """
+    Converts the given date/time periods to date periods.
+
+    The given date converter is used to decide what date to ascribe to a
+    given date/time, since we lose timezone info in this conversion.
+
+    We need to decide what to do with partial days at the start and end:
```

Comment:
🧀 These are heuristics that are (I think) commonly what we want, but may still be surprising - another approach would be to (optionally) error if we have partial days.

Reply:
I think the current behaviour is a bit surprising, since we treat the start and end differently (e.g. if we go from midday to midday, we keep the first day and drop the last).

To keep things simple, I'd be happy to error if we aren't midnight aligned.

Alternatively, I think we could have two functions: `as_covering_date_periods` and `as_covered_date_periods`, which more explicitly cover both behaviours?